app/controllers/history_controller.rb

```diff
@@ -0,0 +1,58 @@
+class HistoryController < LinksControllerBase
+  before_action :authorize, only: %i[update]
+
+  # 2. set @link instance var, since a lot of action filters use it
+  before_action do
+    set_link(params[:link_id])
+  end
+  before_action :require_link
+  before_action only: %i[show update] do
+    set_past_link(params[:id])
+  end
+  before_action :show, only: %i[show]
+
+  # 3. protect link-specific buisness rules
+  before_action :past_link_matches_link, only: %i[show update]
+  before_action :prevent_public_expired, if: -> { link_is_expired_and_not_owned? }
+  before_action :skip_unauthorized_requests, only: %i[update], if: -> { !past_link_belongs_to_user? }
+  before_action :protect_friends_only_links, only: %i[index show update]
+
+  after_action :track_visit, only: %i[index show]
+
+  def index
+    @past_links = PastLink.all.order(id: :desc).where(link: @link).take(50)
+  end
+
+  def show
+  end
+
+  def update
+    @past_link.set_reaction(past_link_params['response_type'], past_link_params['response_text'])
+    unless @past_link.destroyed?
+      redirect_to link_past_link_path(@link, @past_link)
+    else
+      headers['Location'] = link_past_links_path(@link)
+      head :reset_content
```

Comment:
So, this is the only thing here that I'm a bit iffy on, as far as whether there's a better way to do this.

Sending a "hated it" reaction will cause the past_link to be destroyed, so the user needs to be redirected to somewhere or else they'll just get hit with a 404.

Currently I'm doing this by sending an HTTP 205 Reset Content along with a location header that points back to the history page. It's not quite standard, but I felt like it made enough sense given the context. 

On the client, the 205 response is handled in JS by a Stimulus controller. Overall, it works and I don't see any functional problems with it, but you can let me know if there's a more idiomatic way.